packages/orchestration/src/util/stream.ts

```diff
@@ -0,0 +1,351 @@
+  for (const [moduleName, moduleResult] of Object.entries(incoming || {})) {
+    switch (moduleName) {
+      case 'llm':
+        mergedModuleResults[moduleName] = mergeLlmModule(
+          mergedModuleResults[moduleName],
+          moduleResult
+        );
+        break;
+      case 'output_unmasking':
+        mergedModuleResults[moduleName] = mergeLlmChoices(
+          mergedModuleResults[moduleName],
+          moduleResult
+        );
+        break;
+      default:
+        mergedModuleResults[moduleName] = moduleResult;
+    }
+  }
+  return mergedModuleResults;
+}
+
+function mergeLlmModule(
+  existing: LlmModuleResult | undefined,
+  incoming: LLMModuleResultStreaming | undefined
+): LlmModuleResult | undefined {
+  if (!incoming) {
+    return existing;
+  }
+  const mergedModuleResults = {
+    ...incoming,
+    usage: mergeTokenUsage(existing?.usage, incoming.usage),
+    choices: mergeLlmChoices(existing?.choices, incoming?.choices)
+  };
+  return mergedModuleResults;
+}
+
+function mergeTokenUsage(
+  existing:
+    | { prompt_tokens: number; completion_tokens: number; total_tokens: number }
+    | undefined,
+  incoming:
+    | { prompt_tokens: number; completion_tokens: number; total_tokens: number }
+    | undefined
+): { prompt_tokens: number; completion_tokens: number; total_tokens: number } {
+  if (incoming) {
+    logger.debug(`Token usage: ${JSON.stringify(incoming)}`);
+  }
+  return {
+    prompt_tokens: incoming?.prompt_tokens ?? existing?.prompt_tokens ?? 0,
+    completion_tokens:
+      incoming?.completion_tokens ?? existing?.completion_tokens ?? 0,
+    total_tokens: incoming?.total_tokens ?? existing?.total_tokens ?? 0
+  };
+}
+
+function mergeLlmChoices(
+  existing: LlmChoice[] | undefined,
+  incoming: LlmChoiceStreaming[] | undefined
+): LlmChoice[] {
+  const mergedChoices = [...(existing ?? [])];
+  for (const choice of incoming ?? []) {
+    const existingChoice = mergedChoices.find(c => c.index === choice.index);
+    if (existingChoice) {
+      // Merge existing choice with incoming choice
+      existingChoice.finish_reason = handleFinishReason(
+        existingChoice.finish_reason,
+        choice.finish_reason,
+        choice.index
+      );
+      existingChoice.logprobs = mergeLogProbs(
+        existingChoice.logprobs,
+        choice.logprobs
+      );
+      existingChoice.message = mergeMessage(
+        existingChoice.message,
+        choice.delta
+      );
+    } else {
+      // Add new choice
+      mergedChoices.push(transformStreamingChoice(choice));
+    }
+  }
+  return mergedChoices;
+}
+
+function mergeMessage(
+  existing: ResponseChatMessage,
+  incoming: ChatDelta | undefined
+): ResponseChatMessage {
+  if (!incoming) {
+    return existing;
+  }
+  return {
+    role: existing.role,
+    content: existing.content + (incoming.content ?? ''),
+    tool_calls: mergeToolCalls(existing.tool_calls, incoming.tool_calls),
+    refusal: incoming.refusal ?? existing.refusal
```

Comment:
[q] Have you seen refusal in the orchestration service response? Is it easy to trigger with the the `input_filtering` ?